### ISSUE TITLE
boto_secgroup.present intermittently when run as part of an Orchestration state

### ISSUE THREAD

**Issue by MDPolaris** [saltstack/salt#47190]
### Description of Issue/Question
I have a state to create a security group as part of a larger orchestration state. The state fails fairly regularly during the orch run. I have tried to reproduce the issue while running the state manually with state.apply however it has not failed when executed as a standalone state. The result is a security group with the name specified, however the Security Group does not getted tagged with Name: <SG_NAME>. No rules are added to the group either. I reviewed this briefly with @gtmanfred in the slack channel and nothing jumped out as bad with the states. Here is the error produced:

>     nv-salt-syndic:
>     ----------
>               ID: cad_worker_sg
>         Function: boto_secgroup.present
>             Name: nv-wc-creo-stage-syndic-cli-test1-sg
>           Result: False
>          Comment: Attempt 1: Returned a result of "False", with the following comment: " Security group nv-wc-creo-stage-syndic-cli-test1-sg created. nv-wc-creo-stage-syndic-cli-test1-sg security group configuration could not be retrieved. nv-wc-creo-stage-syndic-cli-test1-sg security group configuration could not be retrieved. nv-wc-creo-stage-syndic-cli-test1-sg security group configuration could not be retrieved."
>                   Attempt 2: Returned a result of "False", with the following comment: "An exception occurred in this state: Traceback (most recent call last):
>                     File "/usr/lib/python2.7/site-packages/salt/state.py", line 1837, in call
>                       **cdata['kwargs'])
>                     File "/usr/lib/python2.7/site-packages/salt/loader.py", line 1794, in wrapper
>                       return f(*args, **kwargs)
>                     File "/usr/lib/python2.7/site-packages/salt/states/boto_secgroup.py", line 184, in present
>                       key=key, keyid=keyid, profile=profile)
>                     File "/usr/lib/python2.7/site-packages/salt/states/boto_secgroup.py", line 240, in _security_group_present
>                       profile=profile)
>                     File "/usr/lib/python2.7/site-packages/salt/modules/boto_secgroup.py", line 398, in create
>                       created = conn.create_security_group(name, description, vpc_id)
>                     File "/usr/lib/python2.7/site-packages/boto/ec2/connection.py", line 3019, in create_security_group
>                       SecurityGroup, verb='POST')
>                     File "/usr/lib/python2.7/site-packages/boto/connection.py", line 1208, in get_object
>                       raise self.ResponseError(response.status, response.reason, body)
>                   EC2ResponseError: EC2ResponseError: 400 Bad Request
>                   <?xml version="1.0" encoding="UTF-8"?>
>                   <Response><Errors><Error><Code>InvalidGroup.Duplicate</Code><Message>The security group 'nv-wc-creo-stage-syndic-cli-test1-sg' already exists for VPC 'vpc-ID'</Message></Error></Errors><RequestID>83ff2c4b-8cda-4ab4-aae2-c3215ef3bce5</RequestID></Response>
>                   "
>                   An exception occurred in this state: Traceback (most recent call last):
>                     File "/usr/lib/python2.7/site-packages/salt/state.py", line 1837, in call
>                       **cdata['kwargs'])
>                     File "/usr/lib/python2.7/site-packages/salt/loader.py", line 1794, in wrapper
>                       return f(*args, **kwargs)
>                     File "/usr/lib/python2.7/site-packages/salt/states/boto_secgroup.py", line 184, in present
>                       key=key, keyid=keyid, profile=profile)
>                     File "/usr/lib/python2.7/site-packages/salt/states/boto_secgroup.py", line 240, in _security_group_present
>                       profile=profile)
>                     File "/usr/lib/python2.7/site-packages/salt/modules/boto_secgroup.py", line 398, in create
>                       created = conn.create_security_group(name, description, vpc_id)
>                     File "/usr/lib/python2.7/site-packages/boto/ec2/connection.py", line 3019, in create_security_group
>                       SecurityGroup, verb='POST')
>                     File "/usr/lib/python2.7/site-packages/boto/connection.py", line 1208, in get_object
>                       raise self.ResponseError(response.status, response.reason, body)
>                   EC2ResponseError: EC2ResponseError: 400 Bad Request
>                   <?xml version="1.0" encoding="UTF-8"?>
>                   <Response><Errors><Error><Code>InvalidGroup.Duplicate</Code><Message>The security group 'nv-wc-creo-stage-syndic-cli-test1-sg' already exists for VPC 'vpc-ID'</Message></Error></Errors><RequestID>fd47a316-0a5e-479e-a131-1a39022b9ef4</RequestID></Response>
>          Started: 17:09:24.567103
>         Duration: 31451.466 ms
>          Changes:
> 
>     Summary for nv-salt-syndic
>     ------------
>     Succeeded: 0
>     Failed:    1
>     ------------
>     Total states run:     1
>     Total run time:  31.451 s

### Setup
Here is the rendered section of the Orch state that calls this state:
```
create_cadworker_sg:
  salt.state:
    - tgt: nv-salt-syndic
    - sls:
      - plm.cad_worker_sg
    - concurrent: True
    - pillar:
        sg_name: nv-wc-creo-stage-syndic-cli-test1-sg
        cad_worker_type: creo
        new_environment_id: SVC0002373
    - require:
      - salt: creo_cadworker_start_notify
```

Here is the rendered plm.cad_worker_sg state:
```
cad_worker_sg:
  boto_secgroup.present:
    - name: nv-wc-creo-ptcstg1-sg
    - description: nv-wc-creo-ptcstg1-sg
    - vpc_id: vpc-ID
    - rules:
      - ip_protocol: tcp
        from_port: 20
        to_port: 21
        cidr_ip:
          - 10.20.2.33/32
      - ip_protocol: tcp
        from_port: 601
        to_port: 601
        cidr_ip:
          - 10.20.2.33/32
      - ip_protocol: tcp
        from_port: 990
        to_port: 990
        cidr_ip:
          - 10.20.2.33/32
      - ip_protocol: tcp
        from_port: 5600
        to_port: 5600
        cidr_ip:
          - 10.20.2.33/32
      - ip_protocol: tcp
        from_port: 50000
        to_port: 50100
        cidr_ip:
          - 10.20.2.33/32
    - rules_egress:
      - ip_protocol: all
        from_port: -1
        to_port: -1
        cidr_ip:
          - 0.0.0.0/0
    - tags:
        Name: nv-wc-creo-ptcstg1-sg
        Owner: 'EMAIL'
        Customer: 'Mike Freitas'
    - region: us-east-1
    - retry:
        attempts: 3
        interval: 15
        until: True
```

### Steps to Reproduce Issue
salt-run state.orch orch_new_plm_environment 'pillar={KEY: VALUE}' -l debug

### Versions Report
I have tested on 2 versions with the same results (2017.7.1 and 2017.7.5):

Salt Version:
           Salt: 2017.7.1

Dependency Versions:
           cffi: 1.6.0
       cherrypy: Not Installed
       dateutil: 2.6.1
      docker-py: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.7.2
        libgit2: 0.24.6
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.8
   mysql-python: Not Installed
      pycparser: 2.14
       pycrypto: 2.6.1
   pycryptodome: 3.4.3
         pygit2: 0.24.2
         Python: 2.7.5 (default, Aug  4 2017, 00:39:18)
   python-gnupg: Not Installed
         PyYAML: 3.11
          PyZMQ: 15.3.0
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.1.4

System Versions:
           dist: centos 7.4.1708 Core
         locale: UTF-8
        machine: x86_64
        release: 3.10.0-693.5.2.el7.x86_64
         system: Linux
        version: CentOS Linux 7.4.1708 Core

Salt Version:
           Salt: 2017.7.5

Dependency Versions:
           cffi: Not Installed
       cherrypy: unknown
       dateutil: 1.5
      docker-py: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.7.2
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.8
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
   pycryptodome: 3.4.3
         pygit2: Not Installed
         Python: 2.7.5 (default, Aug  4 2017, 00:39:18)
   python-gnupg: Not Installed
         PyYAML: 3.11
          PyZMQ: 15.3.0
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.1.4

System Versions:
           dist: centos 7.4.1708 Core
         locale: UTF-8
        machine: x86_64
        release: 3.10.0-514.10.2.el7.x86_64
         system: Linux
        version: CentOS Linux 7.4.1708 Core

**Comment by MDPolaris** [comment]
Just another point of interest that could be helpful. I was just doing more testing and hit a string of failures and without any changes they have now resolved. One thing i did notice during the string of failures is that the SG's were slow to delete from the console when i was cleaning them up after each test.

One thought I had was if the AWS API was a little slow in responding and the tag operation fails, which then causes the rest of the operations to fail.

**Comment by MDPolaris** [comment]
Here is another state in the same Orch that is also failing. This was once more reliable and I don't understand why it is failing more often now that the other SG state was added to the Orch.

> [INFO    ] Loading fresh modules for state activity
> [DEBUG   ] LazyLoaded jinja.render
> [DEBUG   ] LazyLoaded yaml.render
> [DEBUG   ] In saltenv 'base', looking at rel_path 'rds_sg.sls' to resolve 'salt://rds_sg.sls'
> [DEBUG   ] In saltenv 'base', ** considering ** path '/var/cache/salt/master/files/base/rds_sg.sls' to resolve 'salt://rds_sg.sls'
> [DEBUG   ] compile template: /var/cache/salt/master/files/base/rds_sg.sls
> [DEBUG   ] Jinja search path: ['/var/cache/salt/master/files/base']
> [PROFILE ] Time (in seconds) to render '/var/cache/salt/master/files/base/rds_sg.sls' using 'jinja' renderer: 0.00107002258301
> [DEBUG   ] Rendered data from file: /var/cache/salt/master/files/base/rds_sg.sls:
> rds_sg:
>   boto_secgroup.present:
>     - name: nv-wc-ptcstg-dev1-rds-sg
>     - description: nv-wc-ptcstg-dev1-rds-sg
>     - vpc_id: vpc-ID
>     - rules:
>       - ip_protocol: tcp
>         from_port: 1521
>         to_port: 1521
>         cidr_ip:
>           - 10.20.128.17/32
>     - rules_egress:
>       - ip_protocol: all
>         from_port: -1
>         to_port: -1
>         cidr_ip:
>           - 0.0.0.0/0
>     - tags:
>         Name: nv-wc-ptcstg-dev1-rds-sg
>         Owner: 'EMAIL'
>         Customer: ptcstg
>         ServiceOfferingNum: SVC0002373
>     - region: us-east-1
> 
> [DEBUG   ] LazyLoaded config.get
> [DEBUG   ] Results of YAML rendering:
> OrderedDict([('rds_sg', OrderedDict([('boto_secgroup.present', [OrderedDict([('name', 'nv-wc-ptcstg-dev1-rds-sg')]), OrderedDict([('description', 'nv-wc-ptcstg-dev1-rds-sg')]), OrderedDict([('vpc_id', 'vpc-ID')]), OrderedDict([('rules', [OrderedDict([('ip_protocol', 'tcp'), ('from_port', 1521), ('to_port', 1521), ('cidr_ip', ['10.20.128.17/32'])])])]), OrderedDict([('rules_egress', [OrderedDict([('ip_protocol', 'all'), ('from_port', -1), ('to_port', -1), ('cidr_ip', ['0.0.0.0/0'])])])]), OrderedDict([('tags', OrderedDict([('Name', 'nv-wc-ptcstg-dev1-rds-sg'), ('Owner', 'EMAIL'), ('Customer', 'ptcstg'), ('ServiceOfferingNum', 'SVC0002373')]))]), OrderedDict([('region', 'us-east-1')])])]))])
> [PROFILE ] Time (in seconds) to render '/var/cache/salt/master/files/base/rds_sg.sls' using 'yaml' renderer: 0.00625014305115
> [DEBUG   ] LazyLoaded boto.assign_funcs
> [DEBUG   ] LazyLoaded boto_secgroup.exists
> [DEBUG   ] LazyLoaded boto_secgroup.present
> [INFO    ] Running state [nv-wc-ptcstg-dev1-rds-sg] at time 21:48:52.132395
> [INFO    ] Executing state boto_secgroup.present for [nv-wc-ptcstg-dev1-rds-sg]
> [DEBUG   ] getting group for nv-wc-ptcstg-dev1-rds-sg in vpc_id vpc-ID
> [INFO    ] Created security group nv-wc-ptcstg-dev1-rds-sg.
> [DEBUG   ] getting group for nv-wc-ptcstg-dev1-rds-sg in vpc_id vpc-ID
> [DEBUG   ] getting group for nv-wc-ptcstg-dev1-rds-sg in vpc_id vpc-ID
> [DEBUG   ] getting group for nv-wc-ptcstg-dev1-rds-sg in vpc_id vpc-ID
> [DEBUG   ] getting group for nv-wc-ptcstg-dev1-rds-sg in vpc_id vpc-ID
> [ERROR   ] {'new': {'secgroup': None}, 'old': {'secgroup': None}}
> [INFO    ] Completed state [nv-wc-ptcstg-dev1-rds-sg] at time 21:48:52.389107 duration_in_ms=256.714
> [DEBUG   ] LazyLoaded event.fire_master
> [WARNING ] Local mode detected. Event with tag salt/job/prog/nv-salt-syndic./0 will NOT be sent.
> [DEBUG   ] File /var/cache/salt/master/accumulator/70598160 does not exist, no need to cleanup.
> [DEBUG   ] LazyLoaded config.option
> [DEBUG   ] Sending event: tag = salt/run/20180419214847984337/ret; data = {'fun_args': ['rds_sg'], 'jid': '20180419214847984337', 'return': {'outputter': 'highstate', 'data': {'nv-salt-syndic': {'boto_secgroup_|-rds_sg_|-nv-wc-ptcstg-dev1-rds-sg_|-present': {'comment': ' Security group nv-wc-ptcstg-dev1-rds-sg created. nv-wc-ptcstg-dev1-rds-sg security group configuration could not be retrieved. nv-wc-ptcstg-dev1-rds-sg security group configuration could not be retrieved. nv-wc-ptcstg-dev1-rds-sg security group configuration could not be retrieved.', 'name': 'nv-wc-ptcstg-dev1-rds-sg', 'start_time': '21:48:52.132393', 'result': False, 'duration': 256.714, '__run_num__': 0, '__sls__': u'rds_sg', 'changes': {'new': {'secgroup': None}, 'old': {'secgroup': None}}, '__id__': 'rds_sg'}}}, 'retcode': 1}, 'success': False, '_stamp': '2018-04-19T21:48:52.392944', 'user': 'root', 'fun': 'runner.state.orch'}
> [DEBUG   ] LazyLoaded highstate.output
> [DEBUG   ] LazyLoaded nested.output
> [TRACE   ] data = {'outputter': 'highstate', 'retcode': 1}
> nv-salt-syndic:
> ----------
>           ID: rds_sg
>     Function: boto_secgroup.present
>         Name: nv-wc-ptcstg-dev1-rds-sg
>       Result: False
>      Comment: Security group nv-wc-ptcstg-dev1-rds-sg created. nv-wc-ptcstg-dev1-rds-sg security group configuration could not be retrieved. nv-wc-ptcstg-dev1-rds-sg security group configuration could not be retrieved. nv-wc-ptcstg-dev1-rds-sg security group configuration could not be retrieved.
>      Started: 21:48:52.132393
>     Duration: 256.714 ms
>      Changes:
>               ----------
>               new:
>                   ----------
>                   secgroup:
>                       None
>               old:
>                   ----------
>                   secgroup:
>                       None
> 
> Summary for nv-salt-syndic
> ------------
> Succeeded: 0 (changed=1)
> Failed:    1
> ------------
> Total states run:     1
> Total run time: 256.714 ms
> [DEBUG   ] LazyLoaded local_cache.prep_jid
> [INFO    ] Runner completed: 20180419214847984337
> [DEBUG   ] Runner return: {'outputter': 'highstate', 'retcode': 1}

**Comment by MDPolaris** [comment]
I would also like to add that I am now seeing these failures while running the states above with state.apply. Originally I did not experience a failure with the states running outside of Orchestration, however i am now experiencing more failures than successes.

**Comment by Ch3LL** [comment]
ping @saltstack/team-boto any ideas here?

**Comment by MDPolaris** [comment]
Hi @Ch3LL I have continued to test and I am starting to lean towards slow responses from the AWS API as the root cause. I have been playing with my retry setting and I did see a run that failed initially but after 5 retires with 30 secs between the state succeeded. I will continue to test, but i wanted to share that in the event the code looks good. At a minimum maybe it is worth a note in the documentation if it turns out to be a timing issue with AWS. Or possible update the error message to suggest adding a retry. That being said I am curious if the boto team has experienced this issue themselves. Thanks.
